### PR TITLE
Conversion to Poetry Build System

### PR DESCRIPTION
**Pull request by kcathey**
Conversion to poetry for build and managing development virtualenv.

### CODE REVIEW

File: README.md
Comment (FrankD412) [inline README.md:167]: This might be a silly question -- should this be done in a virtual environment?

Comment (FrankD412) [reply README.md:167]: Might also be worth linking to the poetry website?

File: pyproject.toml
Comment (FrankD412) [inline pyproject.toml:1]: Another silly question and probably one I should know the answer to -- Pypi official support `toml` now?

**Review by FrankD412 — CHANGES_REQUESTED** [review]

**Comment by FrankD412** [comment]
I just gotta say, liking the looks of Poetry haha

File: README.md
Comment (FrankD412) [reply README.md:167]: From what I saw, this only installed dependencies, it didn't install Maestro itself. Would we still need the `pip install -e .` command?

File: pyproject.toml
Comment (FrankD412) [inline pyproject.toml:17]: We're dropping support for 2.7, so the `chainmap` dependency can go. I'm looking to see how we limit Python versions, but I'm not seeing that currently other than in the `classifiers`.

File: README.md
Comment (FrankD412) [reply README.md:167]: > From what I saw, this only installed dependencies, it didn't install Maestro itself. Would we still need the `pip install -e .` command?

Ok -- I think I verified that this installs.



> This might be a silly question -- should this be done in a virtual environment?

It doesn't look like this should be in a virtualenv at least from my toying with poetry.

Comment (kcathey) [reply README.md:167]: So poetry should be installed system wide. It then creates and manages the virtualenv for you, kind of like pipenv. A `poetry install` will install the project, just like doing a `pip -e .`

File: pyproject.toml
Comment (kcathey) [reply pyproject.toml:1]: Yes indeed. To deploy to pypi all you need to do is, `poetry deploy`. It's kinda awesome.

File: pyproject.toml
Comment (kcathey) [reply pyproject.toml:17]: That's the one piece I haven't worked out as of yet.

File: README.md
Comment (FrankD412) [reply README.md:167]: Yeah, I just figured that out. I'm basically going to configure it to make the environments in my `WORKON_HOME` directory. It might be worth noting that so folks can use virtualenvs for other projects?

File: pyproject.toml
Comment (FrankD412) [reply pyproject.toml:1]: I saw that in their documentation -- this looks like it's gonna be really rad. I'm gonna port pyaestro to using it.

Comment (kcathey) [reply pyproject.toml:1]: I've been using for my django plugins for work, it's really easy to get it to work with gitlab and pypi.

File: README.md
Comment (FrankD412) [reply README.md:167]: Sorry about mucking up the comments also, I was trying to get oriented with the new-ness haha

Might be worth also making a discussion thread about `poetry` best practices that we could port to docs.

Comment (kcathey) [reply README.md:167]: That's what the comments are for.

A discussion thread might indeed be a good idea.

Comment (FrankD412) [reply README.md:167]: I'll start one once this gets merged -- I want to make an introduce yourself post as well. I think it might help with community building!

File: pyproject.toml
Comment (FrankD412) [reply pyproject.toml:17]: I think this is the last bit, I'll approve it for merge once we figure this out.

Comment (FrankD412) [reply pyproject.toml:17]: Oh, actually -- doing some reading, we might need to include data files through the `toml` file too..

Comment (FrankD412) [reply pyproject.toml:17]: > That's the one piece I haven't worked out as of yet.

I found some projects that include a `pyproject.toml` [here](https://github.com/carlosperate/awesome-pyproject#project-templates) and it seems you can specify `python` similar to below.

```
[tool.poetry.dependencies]
python = "^3.7"
importlib_metadata = {version = "^1.6.0", python = "<3.8"}
typer = {extras = ["all"], version = "^0.2.1"}
rich = "^2.2.2"
pre-commit = "^2.5.1"
```

Comment (kcathey) [reply pyproject.toml:17]: I double checked the build and it pulls in the data file.

Comment (kcathey) [reply pyproject.toml:17]: > Oh, actually -- doing some reading, we might need to include data files through the `toml` file too..

I double checked the build and it pulls in the data file.

Comment (kcathey) [reply pyproject.toml:17]: Just to be clear we are supporting python >= 3.5 correct?

Comment (FrankD412) [reply pyproject.toml:17]: Yep -- that's right. I reformatted some stuff last night and add `'>3.4' for the `python` line and it ran super slow in WSL. It looks like it successfully finished in the middle of the night. Here's what I tested:

```
[tool]
[tool.poetry]
name = "maestrowf"
version = "1.1.9.dev0"
description = "A tool to easily orchestrate general computational workflows both locally and on supercomputers."
license = "MIT License"
classifiers = [
    "Development Status :: 5 - Production/Stable",
    "Operating System :: Unix", "Operating System :: MacOS :: MacOS X",
    "Intended Audience :: Developers", "Intended Audience :: Education",
    "Intended Audience :: Science/Research", "Topic :: Scientific/Engineering",
    "Topic :: System :: Distributed Computing",
    "Programming Language :: Python :: 3",
    "Programming Language :: Python :: 3.5",
    "Programming Language :: Python :: 3.6",
    "Programming Language :: Python :: 3.7",
    "Programming Language :: Python :: 3.8",
]
homepage = "https://github.com/llnl/maestrowf"
documentation = "https://maestrowf.readthedocs.io/"
authors = [
    "Francesco Di Natale <dinatale3@llnl.gov>", "Francesco Di Natale"
]
maintainers = [
    "Francesco Di Natale <dinatale3@llnl.gov>", "Francesco Di Natale"
]
readme = "README.md"
packages = [
    { include = 'maestrowf'}
]

[tool.poetry.dependencies]
python = ">3.4"
coloredlogs = "*"
dill = "*"
filelock = "*"
jsonschema = ">=3.2.0"
pyyaml = ">=4.2b1"
six = "*"
tabulate = "*"

[tool.poetry.dev-dependencies]
fabric = "*"
coverage = "*"
sphinx_rtd_theme = "*"
sphinx = "*"
flake8 = "*"
pydocstyle = "*"
pylint = "*"
tox = "*"
pytest = "*"
pytest-cov = "*"
pre-commit = "*"
tox-travis = "*"
tox-pyenv = "*"


[tool.poetry.plugins."console_scripts"]
"maestro" = "maestrowf.maestro:main"
"conductor" = "maestrowf.conductor:main"


[build-system]
requires = ["poetry>=0.12"]
build-backend = "poetry.masonry.api"


[tool.poetry.urls]
"Bug Tracker" = "https://github.com/LLNL/maestrowf/issues"
"Discussions" = "https://github.com/LLNL/maestrowf/discussions"
```

Comment (FrankD412) [reply pyproject.toml]: Relatedly, I listed myself as a maintainer -- feel free to add yourself @kcathey , I was planning to look up the right emails. @jwhite242 @jsemler -- want to be added?

**Review by FrankD412 — APPROVED** [review]

**Comment by FrankD412** [comment]
@kcathey -- The maintainers is the last thing. The only other thing I can think of is if the `readme` key will have the README show up on Pypi. Other than that it looks like this does everything it needs to recreate the current `setuptools` implementation.

**Comment by kcathey** [comment]
> @kcathey -- The maintainers is the last thing. The only other thing I can think of is if the `readme` key will have the README show up on Pypi. Other than that it looks like this does everything it needs to recreate the current `setuptools` implementation.

I've not been able to find the `readme` key documented in the poetry docs, but I've found numerous blog posts stating that it will, and an existing project: flake8-markdown - [Blog](https://johnfraney.ca/posts/2019/05/28/create-publish-python-package-poetry/) [PyPi](https://pypi.org/project/flake8-markdown/) [Project](https://github.com/johnfraney/flake8-markdown)

**Pull request merged**
